neo/Consensus/PrepareRequest.cs

```diff
@@ -12,9 +12,16 @@ internal class PrepareRequest : ConsensusMessage
         public UInt160 NextConsensus;
         public UInt256[] TransactionHashes;
         public MinerTransaction MinerTransaction;
-        public byte[] Signature;
+        /// <summary>
+        /// Prepare Request payload signature
+        /// </summary>
+        public byte[] PrepReqSignature;
+        /// <summary>
+        /// Final block signature
+        /// </summary>
+        public byte[] FinalSignature;
```

Comment:
the primary should not send `FinalSignature` at the beginning. It should be sent in the Commit phase just like backups.

Reply:
Hi @erikzhang,

In the first draft it was done as you suggested.
But later we thought that `FinalSignature` can be sent as a speed-up!
The primary is committed since the beginning since he is the one that proposed the block. With this we gain one signature that we already known. As far as we analyzed there is no negative aspects and makes sense.

Reply:
A node that has sent a `FinalSignature` is not allowed to change view. This implementation violates the design I proposed earlier.

Reply:
In fact, there is other way of looking...aehauheau

A node that is sees the commitment of the majority (has `M` partial signatures`) is not allowed to change view.

I think it is better to send it earlier, Erik, we gain one signature. The Speaker will not send it wrong (furthermore, we verify it) because if he send it wrong everyone will now that he is byzantine.

Reply:
In addition, in the new design every Primary should propose it own set of transactions. Thus, it is almost impossible that we can cheat the system by storing this `FinalSignature`.

Reply:
@vncoelho I agree with Erik on this. Primary should vote again in commit phase, so it should send out his FinalSignature in commit phase .

Reply:
I need to understand better, I can not see any problem or negative aspect in sending the finalsignature in advance (as done today).

In fact, we have been just thinking about the speed up and the logic that primary proposed and, consequently, is already commited.

Reply:
If a node tries to double spend, he will send two transactions, T1 and T2. If the primary receives T1, the other node receives T2, even if the primary is not a byzantine node, it will send a `PrepareRequest` that the backups cannot accept. So even the primary must send the `FinalSignature` after M consensus nodes send the `PrepareResponse`.

Reply:
I still did not get the idea, Erik. Signatures are all reset in case of change_view.

Reply:
The core idea is that primary is not unconditionally accept its own proposal. The criteria for his approval are the same as the backups, that is, there are more than M consensus nodes to accept the proposal. `PrepareRequest` and `PrepareResponse` are the initial judgments of the node on the proposal, and `FinalSignature` is the ultimate judgment of the entire consensus team on the proposal.

Reply:
I agree, @erikzhang.

The `FinalSignature` does not mean that `Primary` accepted his own proposal.
`Primary` will just enter commit phase equally as other Backups, when `M` valid approvals are achieved.

Reply:
send `FinalSignature_{primary}`, in advance, or not send, this is the question. ahueahueahueahueaea

Reply:
Do we have a reached a decision about this? :D 
Or should we really remove the `FinalSignature_{primary}` from being sent in advance?